# Project 1: Data Cleaning & Preparation

## 1. Import Libraries

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## 2. Load Dataset

In [16]:
df = pd.read_excel("../data/raw/Dataset for Data Analytics.xlsx")

## 3. Initial Data Inspection

In [37]:
df.head()


,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
0,ORD200000,2023-01-04,C72649,Monitor,5,570.62,928 Main St,Debit Card,Shipped,TRK37947903,7,SAVE10,Instagram,2853.10
1,ORD200001,2024-08-23,C75739,Phone,2,151.35,823 Main St,Online,Shipped,TRK91186779,3,SAVE10,Referral,302.70
2,ORD200002,2024-02-27,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,FREESHIP,Email,2753.40
3,ORD200003,2023-10-15,C33540,Chair,1,273.19,275 Main St,Debit Card,Returned,TRK62788070,5,SAVE10,Facebook,273.19
4,ORD200004,2025-05-08,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04


In [18]:
df.shape

(1200, 14)

In [19]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   OrderID          1200 non-null   str           
 1   Date             1200 non-null   datetime64[us]
 2   CustomerID       1200 non-null   str           
 3   Product          1200 non-null   str           
 4   Quantity         1200 non-null   int64         
 5   UnitPrice        1200 non-null   float64       
 6   ShippingAddress  1200 non-null   str           
 7   PaymentMethod    1200 non-null   str           
 8   OrderStatus      1200 non-null   str           
 9   TrackingNumber   1200 non-null   str           
 10  ItemsInCart      1200 non-null   int64         
 11  CouponCode       891 non-null    str           
 12  ReferralSource   1200 non-null   str           
 13  TotalPrice       1200 non-null   float64       
dtypes: datetime64[us](1), float64(2), int64(2), str(9)


## 4. Missing Value Analysis


In [20]:
df.isnull().sum()

OrderID              0
Date                 0
CustomerID           0
Product              0
Quantity             0
UnitPrice            0
ShippingAddress      0
PaymentMethod        0
OrderStatus          0
TrackingNumber       0
ItemsInCart          0
CouponCode         309
ReferralSource       0
TotalPrice           0
dtype: int64

## 5. Duplicate Check

In [21]:
df.duplicated().sum()

np.int64(0)

No duplicate rows were found, so no records were removed.

## 6. Data Validation

In [22]:
df["PaymentMethod"].unique()

<StringArray>
['Debit Card', 'Online', 'Credit Card', 'Gift Card', 'Cash']
Length: 5, dtype: str

### Payment Method Validation

The `PaymentMethod` column contains five unique and consistently formatted values. No cleaning was required.

In [23]:
df["OrderStatus"].unique()

<StringArray>
['Shipped', 'Cancelled', 'Returned', 'Delivered', 'Pending']
Length: 5, dtype: str

### Order Status Validation

The `OrderStatus` column contains five valid and consistently formatted categories. No cleaning was required.

In [24]:
df["ReferralSource"].unique()

<StringArray>
['Instagram', 'Referral', 'Email', 'Facebook', 'Google']
Length: 5, dtype: str

### Referral Source Validation

The `ReferralSource` column contains five unique and consistently formatted values. No cleaning was required.

In [25]:
df["CouponCode"] = df["CouponCode"].fillna("No Coupon")

In [26]:
df.isnull().sum()

OrderID            0
Date               0
CustomerID         0
Product            0
Quantity           0
UnitPrice          0
ShippingAddress    0
PaymentMethod      0
OrderStatus        0
TrackingNumber     0
ItemsInCart        0
CouponCode         0
ReferralSource     0
TotalPrice         0
dtype: int64

### Handling Missing Values

The `CouponCode` column contained 309 missing values.

Since a missing coupon code likely indicates that no coupon was applied during the purchase, the missing values were replaced with `"No Coupon"` instead of removing records.

In [28]:
(df["Quantity"] < 0).sum()

np.int64(0)

In [29]:
(df["UnitPrice"] < 0).sum()

np.int64(0)

In [30]:
(df["Quantity"] * df["UnitPrice"] == df["TotalPrice"]).all()

np.False_

In [31]:
mismatch = df[df["Quantity"] * df["UnitPrice"] != df["TotalPrice"]]

mismatch

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
2,ORD200002,2024-02-27,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,FREESHIP,Email,2753.40
10,ORD200010,2023-12-29,C43443,Tablet,5,625.97,333 Main St,Credit Card,Returned,TRK98859248,9,WINTER15,Instagram,3129.85
11,ORD200011,2024-02-17,C93861,Monitor,3,49.14,831 Main St,Online,Returned,TRK48234646,7,SAVE10,Email,147.42
32,ORD200032,2023-05-01,C12388,Tablet,5,536.72,830 Main St,Cash,Delivered,TRK99261395,10,No Coupon,Email,2683.60
41,ORD200041,2023-12-05,C20212,Monitor,3,611.45,657 Main St,Gift Card,Pending,TRK69733327,4,FREESHIP,Instagram,1834.35
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1122,ORD201122,2023-06-07,C38840,Monitor,5,678.19,766 Main St,Online,Returned,TRK32496970,8,No Coupon,Facebook,3390.95
1136,ORD201136,2023-09-24,C71899,Tablet,3,668.33,799 Main St,Gift Card,Pending,TRK34856025,4,WINTER15,Facebook,2004.99
1175,ORD201175,2024-09-10,C62512,Printer,3,91.74,368 Main St,Debit Card,Returned,TRK24139954,8,FREESHIP,Referral,275.22
1189,ORD201189,2024-08-17,C96136,Tablet,5,389.22,149 Main St,Cash,Cancelled,TRK60856682,7,No Coupon,Referral,1946.10


In [32]:
len(mismatch)

107

In [33]:
mismatch[["OrderID", "Quantity", "UnitPrice", "TotalPrice"]]

,OrderID,Quantity,UnitPrice,TotalPrice
2,ORD200002,5,550.68,2753.40
10,ORD200010,5,625.97,3129.85
11,ORD200011,3,49.14,147.42
32,ORD200032,5,536.72,2683.60
41,ORD200041,3,611.45,1834.35
...,...,...,...,...
1122,ORD201122,5,678.19,3390.95
1136,ORD201136,3,668.33,2004.99
1175,ORD201175,3,91.74,275.22
1189,ORD201189,5,389.22,1946.10


In [34]:
(
    (df["Quantity"] * df["UnitPrice"]).round(2)
    == df["TotalPrice"].round(2)
).all()

np.True_

## Total Price Validation

The `TotalPrice` column was validated by comparing it with the calculated value (`Quantity × UnitPrice`).

The initial comparison returned `False` because floating-point numbers are stored with small precision differences in computers.

After rounding both values to two decimal places, all records matched successfully.

**Result:** The `TotalPrice` values are correct.

In [35]:
df.to_excel("../data/cleaned/Cleaned_Dataset.xlsx", index=False)

In [36]:
print("Missing Values:")
print(df.isnull().sum())

print("\nDuplicate Rows:")
print(df.duplicated().sum())

Missing Values:
OrderID            0
Date               0
CustomerID         0
Product            0
Quantity           0
UnitPrice          0
ShippingAddress    0
PaymentMethod      0
OrderStatus        0
TrackingNumber     0
ItemsInCart        0
CouponCode         0
ReferralSource     0
TotalPrice         0
dtype: int64

Duplicate Rows:
0


# Conclusion

The dataset was successfully cleaned and validated.

### Tasks Performed
- Loaded and inspected the dataset.
- Identified and handled missing values.
- Verified duplicate records.
- Validated categorical columns.
- Verified the `TotalPrice` calculation after accounting for floating-point precision.
- Saved the cleaned dataset for further analysis.

The dataset is now ready for data analysis and visualization.